src/dac/dual.rs

```diff
@@ -0,0 +1,192 @@
+#![deny(missing_docs)]
+
+//! Dual DAC
+//!
+//! This module provides access to the dual DAC channels of the STM32G4 series microcontrollers.
+//! It allows for simultaneous access to both DAC channels and supports DMA transfers.
+//!
+
+use core::marker::PhantomData;
+
+use embedded_hal::delay::DelayNs;
+
+use crate::dac::{
+    format::SampleFormat, trigger::DacTriggerSource, DacCh, DacChannel, Disabled, Enabled,
+    Instance, M_EXT_PIN,
+};
+use crate::dac::{hfsel, DacExt as _, Pins};
+
+use crate::dac::format::{self, ToDac as _};
+use crate::dma::mux::DmaMuxResources;
+use crate::dma::traits::TargetAddress;
+use crate::dma::MemoryToPeripheral;
+use crate::rcc::Rcc;
+
+/// Extension trait for [`Instance`] to create a [`DualDac`] from the given pins and RCC.
+pub trait DualDacExt: Instance + Sized {
+    /// Create a dual DAC instance from the given pins and RCC.
+    ///
+    /// DAC calibration is performed before enabling the DAC, so
+    /// a DelayNs implementation is required.
+    ///
+    /// This uses the dual hold registers (DHR12xD, DHR8RD) and
+    /// can write to both DAC channels simultaneously with DMA support.
+    fn into_dual<F: SampleFormat, PINS>(
+        self,
+        pins: PINS,
+        rcc: &mut Rcc,
+        delay: &mut impl DelayNs,
+    ) -> DualDac<Self, F, Disabled>
+    where
+        PINS: Pins<
+            Self,
+            Output = (
+                DacCh<Self, 0, M_EXT_PIN, Disabled>,
+                DacCh<Self, 1, M_EXT_PIN, Disabled>,
+            ),
+        >;
+}
+
+/// Dual DAC handle
+pub struct DualDac<DAC: Instance, F: SampleFormat, State> {
+    _ch1: DacCh<DAC, 0, M_EXT_PIN, State>,
+    _ch2: DacCh<DAC, 1, M_EXT_PIN, State>,
+    _phantom: PhantomData<F>,
+}
+
+impl<DAC: Instance, F: SampleFormat> DualDac<DAC, F, Disabled> {
+    /// Enable DMA double mode for the specified channel.
+    /// This creates a single DMA request for every
+    /// two external hardware triggers (excluding software triggers).
+    #[inline(always)]
+    pub fn enable_dma_double(&mut self, channel: DacChannel, enable: bool) {
+        let dac = unsafe { &(*DAC::ptr()) };
+        dac.mcr()
+            .modify(|_, w| w.dmadouble(channel as u8).bit(enable));
+    }
+
+    /// Enable DMA for the specified channel
+    #[inline(always)]
+    pub fn enable_dma(&mut self, channel: DacChannel, enable: bool) {
```

Comment:
Why does `DacCh::enable_dma` consume self and turn it into a new type while this one does not?

Also, is it ok to have `set_channels` still available after dma has been enabled?